tests/compile-and-dump/Singletons/StandaloneDeriving.hs

```diff
@@ -0,0 +1,32 @@
+module Singletons.StandaloneDeriving where
+
+import Data.Singletons.Prelude
+import Data.Singletons.Prelude.Show
+import Data.Singletons.TH
+
+$(singletons [d|
+
+  infixl 6 :*:
+  data T a = a :*: a
+  data S = S1 | S2 deriving Eq
```

Comment:
OK, after reading the come a little more carefully, it appears that we call `partitionDecs` _before_ `singDataD`/`promoteDataDec`, which handle `Eq`. So I believe what I need to do is:

1. Make it so that `partitionDecs` does _not_ discard standalone deriving declarations it cannot handle, but instead defer them so that `singDataD`/`promoteDataDec` may act on them. I believe I could accomplish this by adding another field to [`PartitionedDecs`](https://github.com/goldfirere/singletons/blob/75dae57bbe06b21b3adbb82158c4767ebd695d14/src/Data/Singletons/Partition.hs#L34) for deferred standalone deriving declarations.
2. In `singDataD`/`promoteDataDec`, check the deferred standalone deriving declarations for `Eq`. If it's there, derive `SEq`/`PEq`/`SDecide`.
3. Otherwise, discard the deferred standalone deriving declarations, since `singletons` can't handle them.

Does that sound agreeable?